src/Bicep.Core/Emit/TemplateWriter.cs

```diff
@@ -515,6 +529,113 @@ private static ObjectExpression GetTypePropertiesForUnionTypeExpression(UnionTyp
             return ExpressionFactory.CreateObject(properties, expression.SourceSyntax);
         }
 
+        private ObjectExpression GetTypePropertiesForDiscriminatedObjectExpression(DiscriminatedObjectTypeExpression expression)
+        {
+            var nullable = false;
+            var nonLiteralTypeName = "object";
+
+            if (TypeHelper.CreateTypeUnion(expression.MemberExpressions.Select(exp => exp.ExpressedType)) is UnionType unionType)
+            {
+                var unionTypeInfo = GetUnionTypeNullabilityAndType(unionType);
+                nullable = unionTypeInfo.Nullable;
+                nonLiteralTypeName = unionTypeInfo.NonLiteralTypeName;
+            }
+
+            var properties = new List<ObjectPropertyExpression>
+            {
+                TypeProperty(nonLiteralTypeName, expression.SourceSyntax),
+                ExpressionFactory.CreateObjectProperty(
+                    "discriminator",
+                    GetTypePropertiesForDiscriminator(expression),
+                    expression.SourceSyntax),
+            };
+
+            if (nullable)
+            {
+                properties.Add(ExpressionFactory.CreateObjectProperty("nullable", ExpressionFactory.CreateBooleanLiteral(true), expression.SourceSyntax));
+            }
```

Comment:
Per the comment in DeclaredTypeManager, it's best to let the wrapper expression take care of this and just treat `DiscriminatedObjectTypeExpression` as representing a non-nullable type. Nullability will be applied by the method handling `NullableTypeExpression`s